core-bundle/src/EventListener/DataContainer/Undo/LabelListener.php

```diff
@@ -0,0 +1,195 @@
+                $labelValues[$k] = $arrRow[$v] ? $this->translator->trans('MSC.yes', [], 'contao_default') : $this->translator->trans('MSC.no', [], 'contao_default');
+                continue;
+            }
+
+            $row_v = StringUtil::deserialize($arrRow[$v]);
+
+            if (\is_array($row_v)) {
+                $args_k = [];
+
+                foreach ($row_v as $option) {
+                    $args_k[] = $GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['reference'][$option] ?: $option;
+                }
+
+                $labelValues[$k] = implode(', ', $args_k);
+                continue;
+            }
+
+            if (isset($GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['reference'][$arrRow[$v]])) {
+                $labelValues[$k] = \is_array($GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['reference'][$arrRow[$v]]) ? $GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['reference'][$arrRow[$v]][0] : $GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['reference'][$arrRow[$v]];
+                continue;
+            }
+
+            if (($GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['eval']['isAssociative'] ?? null) || (ArrayUtil::isAssoc($GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['options'] ?? null) && isset($GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['options'][$arrRow[$v]]))) {
+                $labelValues[$k] = $GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['options'][$arrRow[$v]];
+                continue;
+            }
+
+            $labelValues[$k] = $arrRow[$v];
+        }
+
+        $label = $labelConfig['format'] ? vsprintf($labelConfig['format'], $labelValues) : implode(', ', $labelValues);
+
+        // Remove empty brackets (), [], {}, <> and empty tags from the label
+        $label = preg_replace('/\( *\) ?|\[ *] ?|{ *} ?|< *> ?/', '', $label);
+        $label = preg_replace('/<[^>]+>\s*<\/[^>]+>/', '', $label);
+
+        if (\is_array($labelConfig['label_callback'] ?? null)) {
+            $callable = System::importStatic($labelConfig['label_callback'][0]);
+
+            if (\in_array($mode, [DataContainer::MODE_TREE, DataContainer::MODE_TREE_EXTENDED], true)) {
+                return $callable->{$labelConfig['label_callback'][1]}($arrRow, $label, $dc, '', false, null);
+            }
+
+            return $callable->{$labelConfig['label_callback'][1]}($arrRow, $label, $dc, $labelValues);
+        }
+
+        if (\is_callable($labelConfig['label_callback'] ?? null)) {
+            if (\in_array($mode, [DataContainer::MODE_TREE, DataContainer::MODE_TREE_EXTENDED], true)) {
+                return $labelConfig['label_callback']($arrRow, $label, $dc, '', false, null);
+            }
+
+            return $labelConfig['label_callback']($arrRow, $label, $dc, $labelValues);
+        }
+
+        return $label ?: (string) $arrRow['id'];
```

Comment:
Can't we always return an array? Having to check the type isn't the best DX. :see_no_evil:

Reply:
Depending on the data container mode the labels are either an array (for showing multiple columns) or a single string. See also https://github.com/contao/contao/pull/3644

Reply:
What @fritzmg said. Also, I need the raw column values to render my own table in the preview. I could wrap everything in a function that always returns a string, but then I would render HTML (the table) in the class itself again.

Reply:
If I get this correctly, the Twig template simply renders all availabe values for `preview` side by side. Why can't `preview` always be an array, then? One element is fine as well. Or am I missing something?

Reply:
The difference here is, that with an array it is tabular data (member list, user list, ...). When it is only a string, it is a label like article, page, content element etc. I could always return an array and use `count()`. But let's first wait for #3644 to be implemented as this makes refactoring necessary anyways.